static variables can be declared using non literal, causing c error

### Describe the bug

When declaring static variables, non literals like other variables, arrays, maps, etc. are allowed. This will result in incorrect c code. 

### Reproduction Steps

```v
@[unsafe]
fn foo() {
	b := 23
	static a := b
}

fn main() {
	foo()
}
```

### Expected Behavior

compile error 
something like 'static variable `a` must be declared with literal'

### Current Behavior

```
test.v:8:2: warning: function `foo` must be called from an `unsafe` block
    6 | 
    7 | fn main() {
    8 |     foo()
      |     ~~~~~
    9 | }
test.v:4:9: warning: unused variable: `a`
    2 | fn foo() {
    3 |     b := 23
    4 |     static a := b
      |            ^
    5 | }
    6 |
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/test.01K3Y9K98ZKCTJ1RWHYR3Q147Q.tmp.c:4535: error: initializer element is not constant
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.11

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.11 7831fb0b82263ce0f2ac6795c5e7b12bd5ea3835.2ac3478
|:-------------------|:-------------------
|OS                  |linux, "Void Linux"
|Processor           |12 cpus, 64bit, little endian, AMD Ryzen 5 5600G with Radeon Graphics
|Memory              |5.27GB/15.52GB
|                    |
|V executable        |/home/lars/v/v
|V last modified time|2025-08-30 16:44:56
|                    |
|V home dir          |OK, value: /home/lars/v
|VMODULES            |OK, value: /home/lars/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/lars
|                    |
|Git version         |git version 2.50.1
|V git status        |weekly.2025.06-1106-g2ac34782
|.git/config present |true
|                    |
|cc version          |cc (GCC) 14.2.1 20250405
|gcc version         |gcc (GCC) 14.2.1 20250405
|clang version       |clang version 17.0.6
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 4.0.3 (a9651ff57165f5710bb09a5fe52590fd6ddb72df)
|glibc version       |ldd (GNU libc) 2.41
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -2304,6 +2304,23 @@ pub fn (expr Expr) pos() token.Pos {
 	}
 }
 
+pub fn (expr Expr) is_constant() bool {
+	return match expr {
+		IntegerLiteral, FloatLiteral, BoolLiteral, StringLiteral {
+			true
+		}
+		InfixExpr, CastExpr {
+			true
+		}
+		UnsafeExpr {
+			expr.expr.is_constant()
+		}
+		else {
+			return false
+		}
+	}
+}
+
 pub fn (expr Expr) is_lvalue() bool {
 	return match expr {
 		Ident, CTempVar { true }

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -2309,7 +2309,7 @@ pub fn (expr Expr) is_constant() bool {
 		IntegerLiteral, FloatLiteral, BoolLiteral, StringLiteral {
 			true
 		}
-		InfixExpr, CastExpr {
+		InfixExpr, CastExpr, ArrayInit {
 			true
 		}
 		UnsafeExpr {

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -661,8 +661,13 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 			}
 		}
 		if mut left is ast.Ident && left.info is ast.IdentVar {
-			if left.info.is_static && right_sym.kind == .map {
-				c.error('maps cannot be static', left.pos)
+			if left.info.is_static {
+				if !right.is_constant() {
+					c.error('cannot initialized static variable with non-literal value',
+						left.pos)
+				} else if right_sym.kind == .map {
+					c.error('maps cannot be static', left.pos)
+				}
 			}
 		}
 		// Single side check

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -662,11 +662,11 @@ or use an explicit `unsafe{ a[..] }`, if you do not want a copy of the slice.',
 		}
 		if mut left is ast.Ident && left.info is ast.IdentVar {
 			if left.info.is_static {
-				if !right.is_constant() {
-					c.error('cannot initialized static variable with non-literal value',
-						left.pos)
-				} else if right_sym.kind == .map {
+				if right_sym.kind == .map {
 					c.error('maps cannot be static', left.pos)
+				} else if !right.is_constant() {
+					c.error('cannot initialized static variable with non-constant value',
+						left.pos)
 				}
 			}
 		}

--- a/vlib/v/checker/tests/static_init_err.out
+++ b/vlib/v/checker/tests/static_init_err.out
@@ -1,0 +1,41 @@
+vlib/v/checker/tests/static_init_err.vv:10:2: warning: function `foo` must be called from an `unsafe` block
+    8 | 
+    9 | fn main() {
+   10 |     foo()
+      |     ~~~~~
+   11 | }
+vlib/v/checker/tests/static_init_err.vv:4:9: warning: unused variable: `aa`
+    2 | fn foo() {
+    3 |     b := 23
+    4 |     static aa := b
+      |            ~~
+    5 |     static bb := 1 + 1
+    6 |     static cc := ''.str()
+vlib/v/checker/tests/static_init_err.vv:5:9: warning: unused variable: `bb`
+    3 |     b := 23
+    4 |     static aa := b
+    5 |     static bb := 1 + 1
+      |            ~~
+    6 |     static cc := ''.str()
+    7 | }
+vlib/v/checker/tests/static_init_err.vv:6:9: warning: unused variable: `cc`
+    4 |     static aa := b
+    5 |     static bb := 1 + 1
+    6 |     static cc := ''.str()
+      |            ~~
+    7 | }
+    8 |
+vlib/v/checker/tests/static_init_err.vv:4:9: error: cannot initialized static variable with non-literal value
+    2 | fn foo() {
+    3 |     b := 23
+    4 |     static aa := b
+      |            ~~
+    5 |     static bb := 1 + 1
+    6 |     static cc := ''.str()
+vlib/v/checker/tests/static_init_err.vv:6:9: error: cannot initialized static variable with non-literal value
+    4 |     static aa := b
+    5 |     static bb := 1 + 1
+    6 |     static cc := ''.str()
+      |            ~~
+    7 | }
+    8 |

--- a/vlib/v/checker/tests/static_init_err.out
+++ b/vlib/v/checker/tests/static_init_err.out
@@ -25,14 +25,14 @@ vlib/v/checker/tests/static_init_err.vv:6:9: warning: unused variable: `cc`
       |            ~~
     7 | }
     8 |
-vlib/v/checker/tests/static_init_err.vv:4:9: error: cannot initialized static variable with non-literal value
+vlib/v/checker/tests/static_init_err.vv:4:9: error: cannot initialized static variable with non-constant value
     2 | fn foo() {
     3 |     b := 23
     4 |     static aa := b
       |            ~~
     5 |     static bb := 1 + 1
     6 |     static cc := ''.str()
-vlib/v/checker/tests/static_init_err.vv:6:9: error: cannot initialized static variable with non-literal value
+vlib/v/checker/tests/static_init_err.vv:6:9: error: cannot initialized static variable with non-constant value
     4 |     static aa := b
     5 |     static bb := 1 + 1
     6 |     static cc := ''.str()

--- a/vlib/v/checker/tests/static_init_err.vv
+++ b/vlib/v/checker/tests/static_init_err.vv
@@ -1,0 +1,11 @@
+@[unsafe]
+fn foo() {
+	b := 23
+	static aa := b
+	static bb := 1 + 1
+	static cc := ''.str()
+}
+
+fn main() {
+	foo()
+}

